Have NavigationNudge and NavigationFeedback self-delete when irrelevant

jacob-sewell opened the pull request:
If the navigation feature does not exist, have the NavigationNudge note self-action.

Fixes #7807

_Replace this with a good description of your changes & reasoning._

### Screenshots

### Detailed test instructions:

- Start with a clean WooCommerce site
- Install WooCommerce Admin from this branch and activate it
- Verify that the Navigation checkbox appears on the Settings / Advanced / Features page.

#### NavigationNudge note

- If the 'You now have access to the WooCommerce navigation' note doesn't show up immediately, run the `wc_admin_daily` cron event. I'm using the WP Crontrol plugin.
- Remove the navigation feature (see below)
- Verify that the NavigationNudge note no longer appears (bonus points: check the database and verify that there is no row in `wp_wc_admin_notes` with `name = 'wc-admin-navigation-nudge'`).
- Reenable (but don't activate) the navigation feature: Just undo whatever you did two steps ago.
- Run the `wc_admin_daily` cron event
- Verify that the NavigationNudge note has been recreated

#### NavigationFeedback note

The NavigationFeedback note should be created by the `wc_admin_daily` cron event when the navigation feature is activated (the checkbox is checked on the Settings / Advanced / Features page), so go through the same steps as the NavigationNudge note but activate the navigation feature.

#### One way to remove the navigation feature:

Use the `'woocommerce_admin_features'` filter to remove the `'navigation'` filter like so (you can put this code anywhere it'll be executed--I'm using a custom plugin):

```
add_filter( 'woocommerce_admin_features', function ( $features ) {
	if (!empty($features)) {
		$features = array_filter($features, function ($feature) { return $feature !== 'navigation'; });
	}
	return $features;
} );
```

<!--
Please add your test instructions to `/TESTING-INSTRUCTIONS.md`.
-->

<!--- Please add a Changelog note

Enter a changelog note using the following CLI command `npm run changelogger -- add` and commit changes. --->

jacob-sewell committed:
If the navigation feature does not exist, have the NavigationNudge note self-action.

jacob-sewell commented:
This works in the scenario I've outlined, but I have a couple issues with it.

- As far as I can tell the biggest issue with this in the wild is people who have the woocommerce-admin plugin installed but update their woo core to the point where the admin plugin gets overridden. So this change won't affect them until it makes it all the way into the admin package in core.
- I hope there's a better place to put this check, but the constructor is the only place I found that was being hit on every load. Anybody got any suggestions for a better place to put it?

joshuatf commented on src/Notes/NavigationNudge.php:
```diff
@@ -30,6 +30,9 @@ class NavigationNudge {
 	 * Attach hooks.
 	 */
 	public function __construct() {
+		if ( ! Features::exists( 'navigation' ) ) {
```
We could potentially run this in the `wc_admin_daily` cron so it's not running on every page load.

https://github.com/woocommerce/woocommerce-admin/blob/cf72d3b23ee637701fc075ef5236201ff2573d3e/src/Events.php#L118

The note would still appear for a day or less when they update, but I don't think that's a big deal.

I'm not sure if we should drop a `NavigationNudge::possibly_remove_note()` in here or simply override the `possibly_add_note()` method to check if removal is required.

mattsherman replied:
> But this may be pushing the bounds of this PR. The alternative is to accept that this change won't take effect until after this version of WooCommerce Admin has been packaged in core.

I think that is acceptable. A more comprehensive fix to avoid issues like this going forward should be tackled separately.

mattsherman replied:
> We could potentially run this in the `wc_admin_daily` cron so it's not running on every page load.

Definitely.

> The note would still appear for a day or less when they update, but I don't think that's a big deal.

Agreed. That is reasonable, given our current architecture.

> I'm not sure if we should drop a `NavigationNudge::possibly_remove_note()` in here or simply override the `possibly_add_note()` method to check if removal is required.

mattsherman replied:
> It would be really nice to have some actions we could hook into before deactivating one of the plugins so we can avoid issues like this going forward.

Would this be something more than the [general deactivation hook support in WordPress](https://developer.wordpress.org/reference/functions/register_deactivation_hook/)?

joshuatf replied:
>Would this be something more than the general deactivation hook support in WordPress?

I don't think the plugin actually gets deactivated does it?  It silently sits in the background; we also would want the ability to possibly run something from the unused version prior to returning early in that plugin.

joshuatf replied:
>It would be clearer to have it in a possibly_remove_note(). Removing the note in a method called possibly_add_note() is not expected at all.

I agree this is more clear, though the logic for notes being deleted is typically going to be the opposite for adding or deleting and we'll end up with a lot of duplicated logic.  For example:

```php
function possibly_add_note() {
  if ( ! $has_feature ) {
    return;
  }
}
function possibly_delete_note() {
  if ( $has_feature ) {
    return;
  }
}
```

Having a single function might save some headaches down the road if we plan to add/remove a number of notes.  Naming needs work, but something along the lines of `should_have_note()`.

Might be worth saving for a follow-up.

joshuatf commented on src/Notes/NavigationNudge.php:
```diff
@@ -30,6 +30,9 @@ class NavigationNudge {
 	 */
 	public function __construct() {
+		if ( ! Features::exists( 'navigation' ) ) {
+			self::action_note( '', 'yes' );
```
I think instead of actioning the note (since it actually was not actioned, maybe we could delete the note from the data store.

`Notes::delete_notes_with_name( self::NOTE_NAME );`

mattsherman replied:
If we delete the note, will the note be recreated if Navigation is enabled again in the future, in `possible_add_note()`?

As long as that will work, I agree that it seems more correct to actually delete the note rather than actioning it.

jacob-sewell replied:
I had a version of this that used `self::possibly_delete_note()` but that removed it from the database entirely rather than setting the `is_deleted` flag like I expected so I went back to actioning. I'll look into `Notes::delete_notes_with_name()`.

joshuatf replied:
Setting the `is_deleted` flag is a better idea if we want the note to not show up again, but I would think this note should (re)appear if the feature is available in the currently active version.

joshuatf reviewed (commented):
Nice work figuring out the source of this issue, @jacob-sewell!

>As far as I can tell the biggest issue with this in the wild is people who have the woocommerce-admin plugin installed but update their woo core to the point where the admin plugin gets overridden. So this change won't affect them until it makes it all the way into the admin package in core.

This is a good point and one we've run into a handful of times.  It would be really nice to have some actions we could hook into before deactivating one of the plugins so we can avoid issues like this going forward.

But this may be pushing the bounds of this PR.  The alternative is to accept that this change won't take effect until after this version of WooCommerce Admin has been packaged in core.

mattsherman commented:
> **One way to remove the navigation feature:** Add the following lines to your `woocommerce-admin.php` file right after the namespace declarations:

It is better to put this in a separate my-plugin that you use for development/testing. Ideally, we should never be modifying product source code when testing something.

mattsherman reviewed (dismissed):
Nice start on this, @jacob-sewell !

mattsherman commented:
@jacob-sewell I put the `status: in progress` label on this one to make it clear that it is not ready for formal review and merge yet. That it is a "Draft" also indicates it. 😄

jacob-sewell commented:
> > **One way to remove the navigation feature:** Add the following lines to your `woocommerce-admin.php` file right after the namespace declarations:
> 
> It is better to put this in a separate my-plugin that you use for development/testing. Ideally, we should never be modifying product source code when testing something.

Yeah that's better. I'll change the test instructions.

jacob-sewell commented:
> If we delete the note, will the note be recreated if Navigation is enabled again in the future, in possible_add_note()?
If Navigation is made _available_ again (but not enabled yet, which is the window in which the nudge is relevant) and `possibly_add_note()` gets run, then my understanding is that it wouldn't add the note if it exists in the database but is hidden because of a flag. (Memo to self: check whether `is_deleted` is checked by `get_notes_with_name()`.)

I shied away from the version that resulted in the row being deleted because it doesn't leave an indication of what happened to the note, but if this is the desired behavior I can do that. The thing I don't know is whether upgrading WooCommerce Admin after WooCommerce core has overridden it is the same under the hood as a manual activation, so I don't know whether the user would get the note back. I am pretty sure that manually removing navigation and then putting it back wouldn't result in recreating the note.

joshuatf commented:
>I am pretty sure that manually removing navigation and then putting it back wouldn't result in recreating the note.

If the note is deleted (not soft deleted) then the user would get the note back again.

jacob-sewell committed:
Create delete_if_not_supported method on NavigationNudge note.
If we encounter this situation again we might consider adding a version of this method to NoteTraits.

jacob-sewell committed:
Add possibly_delete_notes() to do_wc_admin_daily().

jacob-sewell commented:
> The thing I don't know is whether upgrading WooCommerce Admin after WooCommerce core has overridden it is the same under the hood as a manual activation, so I don't know whether the user would get the note back.

I do know now. If the NavigationNudge doesn't exist in the database and the conditions are right to create it, it will be created during the `wc_admin_daily` cron event.

> I am pretty sure that manually removing navigation and then putting it back wouldn't result in recreating the note.

As @joshuatf pointed out, this is true of a soft delete but not a hard delete, which is what the code is doing at this point.

jacob-sewell committed:
Merge remote-tracking branch 'origin/main' into fix/7807-navigation-nudge-shows-up-when-invalid

jacob-sewell committed:
Add changelog entry for Issue 7807.

joshuatf commented on src/Events.php:
```diff
@@ -156,6 +157,13 @@ protected function possibly_add_notes() {
+	/**
+	 * Deletes notes that should be deleted.
+	 */
+	protected function possibly_delete_notes() {
```
💯

jacob-sewell replied:
> Having a single function might save some headaches down the road if we plan to add/remove a number of notes. Naming needs work, but something along the lines of `should_have_note()`.

I had considered the idea of something like a `should_note_exist()` method, yeah.

The biggest problem I have with making this feel right (whatever that means) is that the "possibly" in the existing `possibly_delete_note()` methods is just about "Hey maybe there's no note to delete" rather than "Hey maybe the note shouldn't be deleted at this time."

> Setting the `is_deleted` flag is a better idea if we want the note to not show up again, but I would think this note should (re)appear if the feature is available in the currently active version.

I almost always prefer a soft delete to a hard one, but I agree the note should reappear when it becomes relevant again. We could add some logic to `possibly_add_note()` that reverses a soft delete, but since soft delete is what happens when the user manually dismisses the note (I haven't checked that on NavigationNudge yet, but that's what happened on the refund and return note), we definitely don't want to create a zombie note situation.

Come to think of it, we kind of have a zombie note situation now, though it's kind of a long walk to get there. Should I add  logic to `delete_if_not_supported()` to keep soft-deleted notes from being hard-deleted, so they don't get recreated later?

joshuatf replied:
>Come to think of it, we kind of have a zombie note situation now, though it's kind of a long walk to get there. Should I add logic to delete_if_not_supported() to keep soft-deleted notes from being hard-deleted, so they don't get recreated later?

That's a great point.  I'll leave a comment below.

joshuatf commented on src/Events.php:
```diff
@@ -156,6 +157,13 @@ protected function possibly_add_notes() {
+	 * Deletes notes that should be deleted.
+	 */
+	protected function possibly_delete_notes() {
+		NavigationNudge::delete_if_not_supported();
```
We also have the `NavigationFeedback` note that could benefit from removal logic while we're updating this.

jacob-sewell replied:
Is two notes enough to justify putting a version of this method in `NoteTraits`?

joshuatf replied:
Definitely!  I think this would be a useful function.

joshuatf reviewed (commented):
Nice work, @jacob-sewell!  There are some other opportunities here for note logic and also plugin downgrading hooks which would be great to get some follow-up issues around if you have time.

Code is testing well; left a comment about adding deletion logic to the other navigation note that occurs when the navigation feature is enabled.

jacob-sewell committed:
Add should_note_exist() method to NavigationNudge note.

jacob-sewell committed:
Add should_note_exist() and delete_if_not_supported() to NavigationFeedback.

jacob-sewell committed:
Add NavigationFeedback::delete_if_not_supported() to possibly_delete_notes() in daily cron.

jacob-sewell committed:
Fix changelog missing PR number.

jacob-sewell committed:
Update changelog to include navigation feedback note.

joshuatf commented on src/Notes/NavigationNudge.php:
```diff
@@ -33,13 +33,38 @@ public function __construct() {
+	/**
+	 * Should this note exist? (The navigation feature should exist.)
+	 */
+	public static function should_note_exist() {
```
Nice!  This is not a bad name by any means, but there may be opportunity to improve.  I'm not sure I can think of a better one at the moment though.

`is_applicable` maybe?

Not a blocker for this PR; we can always iterate on this.

joshuatf commented on src/Notes/NavigationNudge.php:
```diff
@@ -33,13 +33,38 @@ public function __construct() {
+	/**
+	 * Delete this note if the navigation feature does not exist, unless the note has been soft-deleted already.
+	 */
+	public static function delete_if_not_supported() {
```
Whatever we go with naming-wise above, I think we should mirror here.  `delete_if_not_applicable`.

And I think it would be great to move this to `NoteTraits` as you suggested.

joshuatf commented on src/Notes/NavigationNudge.php:
```diff
@@ -33,13 +33,38 @@ public function __construct() {
+			if ( ! empty( $note_ids ) ) {
+				$note = Notes::get_note( $note_ids[0] );
+
+				if ( ! $note->get_is_deleted() ) {
```
I think we should also check if the note is actioned here as well.  If a note is actioned, we should not resurrect it.

jacob-sewell replied:
Good catch! I will make that change.

joshuatf reviewed (commented):
Great thoughts on this, @jacob-sewell!  Left a few more comments, but the changes here are great and a big improvement to WCA notes.

jacob-sewell committed:
Rename should_note_exist and delete_if_not_supported to is_applicable and delete_if_not_applicable, add to NoteTraits.

joshuatf commented on src/Events.php:
```diff
@@ -156,6 +157,14 @@ protected function possibly_add_notes() {
+	 */
+	protected function possibly_delete_notes() {
+		NavigationNudge::delete_if_not_applicable();
+		NavigationFeedback::delete_if_not_applicable();
```
One more I completely forgot about that could benefit from this!  `NavigationFeedbackFollowUp`

jacob-sewell replied:
I'm creating an issue for `NavigationFeedbackFollowUp`.

joshuatf reviewed (approved):
This is a terrific abstraction for notes; thanks, @jacob-sewell!

Left one more note that could benefit, but optional and could always be done in a follow up. Pre-approving ✅
